### PR TITLE
Capability to specify mount propagation mode of per volume with docker

### PR DESCRIPTION
**Pull request by m3ngyang**
Volume mount propagation mode has been supported since docker 1.10.0, but this function is absent in current kubelet.
This patch set allows one to specify mount propagation setting of a volume. Allowed values are "shared" or "slave" or "private", via introducing a new filed called `"Propagation"` in type `VolumeMount` in `api/types`

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/20698)

### CODE REVIEW

**Comment by Kaffa-MY** [comment]
@k8s-bot ok to test

**Comment by thockin** [comment]
What is a user going to do with this flag?  Mount prop. flags are INCREDIBLY complicated - what behavior do people need that this satisfies?

**Comment by Kaffa-MY** [comment]
Our team attempts to containerize flocker with zfs as back-end storage, and containers in the same flocker node need to read/write and share the same mounted volume. Without this flag, the volume mount propagation mode cannot be specified between the host and the container,  and then the volume mount of each container would be isolated by namespace from each other.
By default, the flag would be set with empty string and parsed as 'rprivate' in docker.
If someone chooses to use the `propagation` flag, he should be familiar with shard mount and responsible for the result himself.

**Comment by thockin** [comment]
@pmorie @swagiaal or whomever is looking at containerized kubelet now that mount.prop is available in Docker.  Is the plan to enable shared by default?  I don't like this as a facet of our API if we can avoid it.

**Comment by Kaffa-MY** [comment]
@thockin besides shared mode, Docker 1.10.0 also support slave and private mode.
fyi [https://github.com/docker/docker/pull/17034](url)

**Comment by thockin** [comment]
I know what it supports, we've been watching it for a long time :)

My point is that this is a REALLY complicated API feature that 0.001% of
users will ever need, and I'd rather DTRT by default than offer a flag.

**Comment by pmorie** [comment]
@Kaffa-MY I agree with @thockin that we need to be careful about exposing an API surface for this.  Currently the default is private in raw docker.

@Kaffa-MY I'm curious about what your use-case for this is.

**Comment by swagiaal** [comment]
Hmm.. I wonder what would be a good way to expose this if not through the API.

Atomic Host has the same use case which is basically containerizing the mount tools so they don't have to be installed on the minimal host. 

Is there a way to make this available to volume plugins without putting it in the API ? That way the change can be made to the flocker plugin as I don't think there is a general use case of pod's needing to propagate mounts.

**Comment by dchen1107** [comment]
cc/ me

**Comment by pmorie** [comment]
So here's one use-case I can imagine: say that we want to containerize mount operations and we want to run those containers in pods.

**Comment by thockin** [comment]
Isn't that what Sami said?

I thought the plan was to move all volumes to shared by default
specifically to allow mount helpers, FUSE daemons, and containerized kubelet

**Comment by pmorie** [comment]
@kubernetes/rh-storage

**Comment by pmorie** [comment]
It is, @thockin, I hadn't seen sami's comment load in the browser window I had this in.

**Comment by Kaffa-MY** [comment]
@pmorie I'm containerizing flocker with zfs as backend storage, containers in the same pod, which constitute a whole flocker agent,  need to share the same volume and the mount information.
@thockin If the `shared` were set as the default propagation mode of volume mount and the API were not exposed, how to deal with the situation where isolation is needed?

**Comment by pmorie** [comment]
@Kaffa-MY  @thockin 

The default propagation mode in docker is private.

**Comment by childsb** [comment]
It sounds like the only scenario we need to support shared mount namespace is in the volume plugins and no need to expose the full options through API. 

So if docker version > 1.10 and mount is volume plugin mount always enable.

**Comment by pmorie** [comment]
I spent a lot of time thinking about this subject this weekend.  Here's one thing that I wonder: do we really need this (does it even really make sense) for anything but hostPath?

**Comment by lvlv** [comment]
I like this change :-)
But it seems pending for a while, do we have a conclusion now? I'm wondering which kind of input could move this forward.

**Comment by childsb** [comment]
@lvlv i'm working on a containerized volume mount proposal. i will link to this PR when ready (should be by the end of week)

**Comment by thockin** [comment]
This is not going to make it for 1.2 release, so I admit that it will be a
bit delayed in getting attention :)

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -391,7 +392,14 @@ func makeMountBindings(mounts []kubecontainer.Mount, podHasSELinuxLabel bool) (r
...
 				bind += ":Z"
 			}
-
+		}
+		// Propagation Mode
+		if len(m.Propagation) != 0 {
+			if !m.ReadOnly && !(m.SELinuxRelabel && podHasSELinuxLabel) {
+				bind += ":" + m.Propagation
+			} else {
+				bind += "," + m.Propagation
```
Comment (rootfs) [inline pkg/kubelet/dockertools/manager.go]: does it work with <1.10 docker?

Comment (Kaffa-MY) [inline pkg/kubelet/dockertools/manager.go]: @rootfs no, docker supports propagation mode from version 1.10. Does k8s need to check the docker version?

Comment (rootfs) [inline pkg/kubelet/dockertools/manager.go]: @Kaffa-MY that's my concern. We better ensure the flag is supported before using it.

**Comment by swagiaal** [comment]
@Kaffa-MY just so you have some feedback for the time being. I think there is general consensus that this be not be exposed in the general API and implemented at the volume plugin level (because that is the only use case we can think of).

Given that I think that the plugin will need to launch a pod (which has the mount tools), make sure the pod is scheduled on the same node where the plugin wants to do the mount and make sure that the container within the pod gets the correct mount propagation setting. Another way is to just directly launch a Docker container which is configured correctly but that removes the runtime abstraction which Kubernetes provides. Just some thoughts..

**Comment by lvlv** [comment]
@swagiaal
The way to launch a pod every time a volume is requested is sort of fragile, the mounter pod may get pending, may run out of control and we don't have a good way to deal with that. We are forced to deal with a lot of dirty work such as cleaning failed / hanging mounter pods.
Instead, we could use DaemonSet to start a daemon on each node, and plugin could call that daemon to do the heavy lifts. In this way, we have more control about the process. As a bonus, scheduler could keep the daemons in mind and assign pod (need that volume) to nodes with daemons.

**Comment by swagiaal** [comment]
> DaemonSet to start a daemon on each node, and plugin could call that daemon to do the heavy lifts

Hmm.. interesting.. would the pod in the daemon set contain the mount tools for all (containerized) plugins ? And how would we set the mount propagation mode for containers in those pods without using the API ?

**Comment by lvlv** [comment]
> would the pod in the daemon set contain the mount tools for all (containerized) plugins ?

We can create different daemons for different mount tools, different plugins could call different daemons on the same host.

> And how would we set the mount propagation mode for containers in those pods without using the API ?

That's why I really want this change :+1: We need mount propagation mode in API so this (a DaemonSet for each customized volume plugin) way becomes feasible.

**Comment by thockin** [comment]
Even if we have something like this in the API, I do not expect it to look
like this.  I would expect it to be higher level and less free.
 `volumeMount.shared` or something.

**Comment by Kaffa-MY** [comment]
If we rewrite the codes and provide the propagation mode API in a more constrained way, does it have any possibility to be accepted? @thockin

**Comment by victorgp** [comment]
Another use case is when using CoreOS that does not provides external fuse systems like, in our case, GlusterFS, and you need a container to do the mounts. The only way to see those mounts in the host,  hence also visible by other containers, is by sharing the mount propagation.

**Comment by marun** [comment]
Similarly to containerized kubelet, I've tested docker 1.10's mount propagation to enable EmptyDir support for docker-in-docker cluster deployment of openshift:

https://github.com/openshift/origin/pull/7668

For this use case, it was sufficient to ensure the docker daemon's default mount propagation was shared since the container root was where the EmptyDir volumes were being mounted.

**Comment by rootfs** [comment]
High level api like `volumeMount.shared` is preferred, this gives us the ability to override default `slave` mode, as needed by #22216 @childsb

**Comment by Kaffa-MY** [comment]
@rootfs api supporting shared propagation mode has been modified to `VolumeMount.Shared`, this is much more constrained. Do you have any other advice to improve the pr and make it prone to be merged?

**Comment by thockin** [comment]
Please let's not merge this until we have an understanding of what we want to do with containerized mount tools.

**Comment by thockin** [comment]
(but this change is less scary now)

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -439,6 +440,18 @@ func makeMountBindings(mounts []kubecontainer.Mount, podHasSELinuxLabel bool) (r
...
+		// Shared Propagation Mode
+		if m.Shared {
+			if supportSharedPropagationMode {
+				if !m.ReadOnly && !(m.SELinuxRelabel && podHasSELinuxLabel) {
+					bind += ":shared"
+				} else {
+					bind += ",shared"
+				}
+			} else {
+				glog.Warningf("Shared propagation mode is to be supported on docker versions >= v1.10")
```
Comment (marun) [inline pkg/kubelet/dockertools/manager.go]: Shared propagation mode requires docker >= v1.10

File: pkg/kubelet/container/runtime.go [hunk 3]
```diff
@@ -326,6 +326,8 @@ type Mount struct {
 	ReadOnly bool
 	// Whether the mount needs SELinux relabeling
 	SELinuxRelabel bool
+	// Whethre the mount is in a shared propagation mode
```
Comment (marun) [inline pkg/kubelet/container/runtime.go]: Whethre -> Whether

File: pkg/api/v1/types.go [hunk 4]
```diff
@@ -869,6 +869,8 @@ type VolumeMount struct {
 	// Path within the container at which the volume should be mounted.  Must
 	// not contain ':'.
 	MountPath string `json:"mountPath"`
+	// Optional: Defaults to false (in a private propagation mode)
```
Comment (marun) [inline pkg/api/v1/types.go]: Consider supplying a more descriptive comment, e.g.

// Use 'shared' mount propagation if true, private otherwise.  Defaults to false.

Comment (Kaffa-MY) [inline pkg/api/v1/types.go]: thanks for your careful review. I'm rewriting now.

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (Kaffa-MY) [inline pkg/kubelet/dockertools/manager.go]: @marun docker version has been checked and result is passed via invariable `supportSharedPropagationMode`

Comment (marun) [inline pkg/kubelet/dockertools/manager.go]: Sorry if I wasn't clear, my intention was to suggest rewording the log message to 'requires docker >= v1.10' instead of 'is to be supported on docker versions >= v1.10'.

Comment (Kaffa-MY) [inline pkg/kubelet/dockertools/manager.go]: @marun sorry, I misunderstood the comment.

**Comment by rootfs** [comment]
@Kaffa-MY thanks for updating. API change and docker version check look good.

**Comment by pmorie** [comment]
I feel very, very strongly that mount propagation controls need to be _only_ on hostpath volumes, and not on `VolumeMount`.

**Comment by Kaffa-MY** [comment]
@pmorie That's right, propagation mode only makes sense when used on hostpath.

**Comment by pmorie** [comment]
@Kaffa-MY that makes it seem like it should be a property of the host volume, not the mountpoint, then.

**Comment by Kaffa-MY** [comment]
What about modifying struct `Attributes` in `pkg/volume/volume.go` to<pre><code>// Attributes represents the attributes of this builder.
type Attributes struct {
	ReadOnly        bool
	Managed         bool
	SupportsSELinux bool
	Shared          bool  // set propagation mode
}</code></pre>Any volume type returns `false` when calling `builder.GetAttributes().Shared`, except for `HostPath`.

**Comment by pmorie** [comment]
There are a couple different aspects of how to do this correctly:

1. What is the right way to expose this in the API
2. What is the right way to implement that API within the kubelet.

The volume attributes struct is strictly related only to (2), let's agree
on an API first.

**Comment by Kaffa-MY** [comment]
First, I plan to expose the api in <pre><code>type HostPathVolumeSource struct {
	Path string `json:"path"`
	Shared bool `json:"shared"` // propagation mode
}</code></pre>As for the implementation of the api within the kubelet, extend the struct `Attributes` to provide a way to access the propagation mode of hostpath volume. Maybe I didn't explain it clearly. :) @pmorie

**Comment by lvlv** [comment]
Hi @pmorie @thockin, this version @Kaffa-MY proposed also works on my case, do you feel more comfortable about it?

> type HostPathVolumeSource struct {
    Path string json:"path"
    Shared bool json:"shared" // propagation mode
}

File: pkg/api/v1/types.go [hunk 5]
```diff
@@ -869,6 +869,8 @@ type VolumeMount struct {
...
+	Shared bool `json:"shared,omitempty"`
```
Comment (mikedanese) [inline pkg/api/v1/types.go]: There are 8 propagation modes here. Why would we only support 2?

**Comment by pmorie** [comment]
@lvlv @Kaffa-MY I feel more comfortable with the idea of the propagation mode being a property of the `HostMountVolumeSource`, but as @mikedanese points out, a boolean for 'shared vs. not' is not the greatest way to model this.  We should have an enum for `PropagationMode` and express each of the possible modes.  Shared vs. 'not-shared' doesn't express the right thing, IMO.

**Comment by Kaffa-MY** [comment]
According to @pmorie 's suggestion, I modified the implementation of propagation mode. Now, `Propagation` is an attribute of volume type `HostPathVolumeSource`. Of course, docker version checking and validation for provided propagation mode are necessary. To access the field, an interface `GetPropagationMode()` is added to the abstract volume type and every concrete volume type needs to implement it. The method returns propagation mode for `HostPath` but empty string for other volume types. Before kubelet mounts the volume, the `Propagation` will be passed to struct `Mount` via interface `GetPropagationMode()` when needed. What about this solution? @thockin

**Comment by kfox1111** [comment]
We could really use this sort of thing. we have an automount based file system we'd like to be able to expose from the host into containers, and the only way I know how to make that work properly is with this sort of feature. We've been doing some bind mounts just into the subdirectories, and it kind of works, but has some very strange side effects.

**Comment by liyimeng** [comment]
Love this! Could not wait for it to be merged!

**Comment by liyimeng** [comment]
@Kaffa-MY are you going on this, or you have given up?

**Comment by Kaffa-MY** [comment]
@liyimeng I'm not going to give up the patch and still waiting for response from k8s members. Besides, I've completed this in our company's k8s.

**Comment by kfox1111** [comment]
They are busy with getting 1.3 out the door. hopefully it is looked at again once that's done. I really could use this too.

File: pkg/volume/volume.go [hunk 6]
```diff
@@ -33,6 +33,9 @@ type Volume interface {
 
 	// MetricsProvider embeds methods for exposing metrics (e.g. used,available space).
 	MetricsProvider
+
+	// GetPropagationMode returns the propagation mode the volume is mounted to.
+	GetPropagationMode() api.PropagationMode
```
Comment (liyimeng) [inline pkg/volume/volume.go]: Why do we need this? Can propagation parameter be read on Attributes? 
Introducing this interface cause update in every volume source, which I believe should be avoided.

**Comment by liyimeng** [comment]
@Kaffa-MY, Great! However I have an question on GetPropagationMode() interface, 
Why do we need this? Can propagation parameter be read on Attributes? 
Introducing this interface cause update in every volume source, which I believe should be avoided.

File: pkg/volume/volume.go (hunk 6 above)
Comment (Kaffa-MY) [inline pkg/volume/volume.go]: @liyimeng we're still talking about how to export the api in a right way. This method is just my personal proposal. If we put propagation mode in the `Attributes`, we also need to update the method `GetAttributes()` in every volume source. And thanks for your advice. :)

Comment (liyimeng) [inline pkg/volume/volume.go]: Thanks @Kaffa-MY!  
Yes, there some issues to sort out here. To understanding, there are two propsals now: 
  spec.containers.volumeMounts.propagation
  spec.containers.volumes.hostPath.propagation
And you are favoring of the second one. If we put propagation mode in the Attributes, Can we just update the method GetAttributes() in hostPath.go? Because other volume source dose not care for this value.

**Comment by victorgp** [comment]
Any chance this will be included in Kubernetes 1.3?

I'm looking forward to seeing this in k8s, many of us use CoreOS and GlusterFS or similar and the lack of this feature in Kubernetes is pushing us back to use it.

**Comment by mikedanese** [comment]
This is not going to make 1.3 but this feature is definitely useful and easy to implement (for docker at least). The hard part is getting the API right which requires some effort and bandwidth to review. @pmor

**Comment by eparis** [comment]
ok to test

**Comment by kfox1111** [comment]
How close do we think this is? I'm starting to contemplate patching this into my own system because I need this sort of functionality.

**Comment by thockin** [comment]
I am out this week, but I agree this has languished.  I'll try to take a
look ASAP and see if we can un-stick it.

**Comment by kfox1111** [comment]
@thockin thanks

I've got a workaround for now. Say you want to share out /export/foo and make it remountable:
I made a systemd unit that does the following after mounts happen but before docker starts:
mount --bind --make-shared /export /export

Then I do a hostPath volume passing through /export. Mounting anything then in /export on the host makes it show up in the container.

But, This patch is still needed. I tried doing the same trick with launching a deamonset with a fuse file system in it, that mounts into the shared /export. It mounts ok and is visible in that container, but not other containers that have /export mounted. so its acting like slave mount, not a shared one. We really would like to containerize this fuse file system, as it has a lot of dependencies and difficult config, so it would be great for kubernetes to manage it.

Maybe this shared flag should be available for adding to emptyDir volumes too? Then you could load fuse file systems as sidecars in a pod.

**Comment by thockin** [comment]
I still do not think the right API is to expose this in all it's detail.  It's too complex and too little value.

ref https://github.com/coreos/rkt/issues/1149

What is the simplest thing we can do that enables the most users?

What if we made all mounts be shared?  Would that enable the host-path case and the kubelet-in-container case?

What if we simply offered Private and Shared and nothing else?  Can we make this part of `type Volume` or `type VolumeMount` rather than per-volume driver?

**Comment by luxas** [comment]
This would be VERY valuable to have in v1.4, given all the @kubernetes/sig-cluster-lifecycle work.
It would make it possible to run kubelet as a pod easily, where `/var/lib/kubelet` has to be a shared mount.

**Comment by luxas** [comment]
@thockin @pmorie @rootfs Any chance you'd pick it up?
It should be documented it only works with docker; but on the other hand, if we some day want to support rkt fly; it won't be supported by docker.

**Comment by andrewmchen** [comment]
+1. Would also really appreciate this feature.

**Comment by andrewmchen** [comment]
@Kaffa-MY Is the full PR including the implementation public?

**Comment by vishh** [comment]
+1 for @thockin's comments. Why can't hostpath volumes be mounted as shared by default? Assuming that not all containers will have the privilege to mount filesystems, are there drawbacks to mounting shared by default?

@luxas Regarding "It would make it possible to run kubelet as a pod easily"

Why is kubelet being run as a pod? I don't think this is an operating mode that @kubernetes/sig-node is willing to support.

**Comment by luxas** [comment]
@vishh Check out self-hosting. It makes a lot of sense.
And running kubelet in a container with a shared mount works really well.
Kubelet in container w/ shared mount != --containerized, which have had problems. 

See https://github.com/kubernetes/kube-deploy/pull/189 for ref.
Most of the fails there was due to the version skew (v1.3.4 vs e2e.test v1.4.0 alpha.2)

I guess the main drawback with default shared mounta is that it will be a docker-specific behaviour then. Otherwise, we could leave it as a docker-specific option that user that understand the implications (and the need for --privileged for it to work) could turn on. Perfect for storafe drivers for example.

We might even be able to containerize the storage driver itself, and then kubelet just could pull and use a such Pod for mounting volumes (like we're using overlay network plugins in a daemonset today). cc @pmorie

**Comment by kfox1111** [comment]
Yeah. I plan on using it for making a daemonset with cvmfs in it instead of flexvolumes.

**Comment by porridge** [comment]
+1 to https://github.com/kubernetes/kubernetes/pull/20698#issuecomment-234720829 - this would make it possible to use Cloud SQL proxy's ``-fuse`` feature.

**Comment by Kaffa-MY** [comment]
@andrewmchen not including implementation, this is just an api proposal

File: pkg/api/types.go [hunk 7]
```diff
@@ -403,10 +403,22 @@ const (
...
+
+const (
+	PropagationPrivate  PropagationMode = "private"
+	PropagationRPrivate PropagationMode = "rprivate"
+	PropagationSlave    PropagationMode = "slave"
+	PropagationRSlave   PropagationMode = "rslave"
+	PropagationShared   PropagationMode = "shared"
+	PropagationRShared  PropagationMode = "rshared"
+)
+
```
Comment (Kaffa-MY) [inline pkg/api/types.go]: @pmorie is this emun ok?

Comment (euank) [inline pkg/api/types.go]: I think based on the discussion here, the next step is to figure out the answer to the following two questions:

1) Can this be solved by a single sane default for volume mounts?
2) If not 1, what's the minimal feature we need to expose to users to enable people the most, and what's a thinner API for it.

It's possible that this *is* the correct minimal thing to expose, but based on the discussion here, it needs to justify itself (and I personally think it is not).

Based on @thockin's comment and concerns (which I share), I think the next step here is to write a proposal explaining what the options are, what the use cases are, and explaining exactly why a given option is the right one.
I think iterating on a .md proposal will be more productive than iterating on the exact code / api-changes.

**Comment by thockin** [comment]
Either this or #31504 has to be closed.  Given that #31504 is more limited in scope, I vote to close this one (no offense).  Scream if I got it wrong  (typical :)
